src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs

```diff
@@ -36,47 +33,42 @@ internal sealed partial class Analysis
             /// and a block that contains the lambda that lifts said variables.
             /// If such block itself requires a closure, then it must lift parent frame pointer into the closure
             /// in addition to whatever else needs to be lifted.
-            /// 
-            /// NOTE: This information is computed in addition to the regular analysis of the tree and only needed for rewriting.
-            /// If someone only needs diagnostics or information about captures, this information is not necessary.
             /// <see cref="ComputeLambdaScopesAndFrameCaptures"/> needs to be called to compute this.
             /// </summary>
-            public HashSet<BoundNode> NeedsParentFrame;
+            public readonly PooledHashSet<BoundNode> NeedsParentFrame = PooledHashSet<BoundNode>.GetInstance();
 
             /// <summary>
             /// Optimized locations of lambdas. 
             /// 
             /// Lambda does not need to be placed in a frame that corresponds to its lexical scope if lambda does not reference any local state in that scope.
             /// It is advantageous to place lambdas higher in the scope tree, ideally in the innermost scope of all scopes that contain variables captured by a given lambda.
             /// Doing so reduces indirections needed when captured locals are accessed. For example locals from the innermost scope can be accessed with no indirection at all.
-            /// 
-            /// NOTE: This information is computed in addition to the regular analysis of the tree and only needed for rewriting.
-            /// If someone only needs diagnostics or information about captures, this information is not necessary.
             /// <see cref="ComputeLambdaScopesAndFrameCaptures"/> needs to be called to compute this.
             /// </summary>
-            public Dictionary<MethodSymbol, BoundNode> LambdaScopes;
+            public readonly SmallDictionary<MethodSymbol, BoundNode> LambdaScopes =
+                new SmallDictionary<MethodSymbol, BoundNode>(ReferenceEqualityComparer.Instance);
 
             /// <summary>
             /// The root of the scope tree for this method.
             /// </summary>
             public readonly Scope ScopeTree;
 
-            private Analysis(Scope scopeTree, HashSet<MethodSymbol> methodsConvertedToDelegates)
+            private Analysis(Scope scopeTree, PooledHashSet<MethodSymbol> methodsConvertedToDelegates)
             {
                 ScopeTree = scopeTree;
                 MethodsConvertedToDelegates = methodsConvertedToDelegates;
             }
 
             public static Analysis Analyze(BoundNode node, MethodSymbol method, DiagnosticBag diagnostics)
             {
-                var methodsConvertedToDelegates = new HashSet<MethodSymbol>();
+                var methodsConvertedToDelegates = PooledHashSet<MethodSymbol>.GetInstance();
                 var scopeTree = ScopeTreeBuilder.Build(
```

Comment:
Consider: would the ownership of `PooledHashSet` here be clearer if `Build` returned it as an `out` parameter? Today it's ambigious at the call site who's responsibility it is to free this value: Build or Analysis.

Also does `Build` assert that the set is empty when passed in? Should it?